fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/ValueIndexExpansionVisitor.java

```diff
@@ -156,13 +156,16 @@ public MatchCandidate expand(@Nonnull final Supplier<Quantifier.ForEach> baseQua
      */
     @Nonnull
     public static KeyExpression fullKey(@Nonnull Index index, @Nullable final KeyExpression primaryKey) {
+        final KeyExpression rootExpression = index.getRootExpression() instanceof KeyWithValueExpression
+                                             ? ((KeyWithValueExpression)index.getRootExpression()).getKeyExpression()
+                                             : index.getRootExpression();
         if (primaryKey == null) {
-            return index.getRootExpression();
+            return rootExpression;
         }
         final var trimmedPrimaryKeyComponents = new ArrayList<>(primaryKey.normalizeKeyForPositions());
         index.trimPrimaryKey(trimmedPrimaryKeyComponents);
         final var fullKeyListBuilder = ImmutableList.<KeyExpression>builder();
-        fullKeyListBuilder.add(index.getRootExpression());
+        fullKeyListBuilder.add(rootExpression);
```

Comment:
This is wrong (and it used to be wrong). We should not just blindly use the `rootExpression` here either in the presence of `KeyWithValue`. I think we should also call `normalizeKeyForPositions()` on the `rootExpression` and then properly pick and merge the pieces from normalized root and from normalized primary..

Reply:
Do you mind changing that? It's great you found this (admittedly quite old) problem...

Reply:
It looks like there was some misunderstanding, after some discussion we found the fix is actually correct 😄 .